episodes/hpc.Rmd

```diff
@@ -0,0 +1,295 @@
+  )
+)
+```
+
+There are a number of options you can pass to `crew_controller_slurm()` to fine-tune the Slurm execution, [which you can find here](https://wlandau.github.io/crew.cluster/reference/crew_controller_slurm.html).
+Here we are only using two:
+
+  * `workers` sets the number of jobs that are submitted to Slurm to process targets.
+  * `script_lines` adds some lines to the Slurm submit script used by Targets. This is useful for loading Environment Modules and adding `#SBATCH` options.
+
+Let's run the modified workflow:
+
+```{R, eval=FALSE}
+source("R/packages.R")
+source("R/functions.R")
+
+library(crew.cluster)
+tar_option_set(
+  controller = crew_controller_slurm(
+    workers = 3,
+    script_lines = "module load R"
+  )
+)
+
+tar_plan(
+  # Load raw data
+  tar_file_read(
+    penguins_data_raw,
+    path_to_file("penguins_raw.csv"),
+    read_csv(!!.x, show_col_types = FALSE)
+  ),
+  # Clean data
+  penguins_data = clean_penguin_data(penguins_data_raw),
+  # Build models
+  models = list(
+    combined_model = lm(
+      bill_depth_mm ~ bill_length_mm, data = penguins_data),
+    species_model = lm(
+      bill_depth_mm ~ bill_length_mm + species, data = penguins_data),
+    interaction_model = lm(
+      bill_depth_mm ~ bill_length_mm * species, data = penguins_data)
+  ),
+  # Get model summaries
+  tar_target(
+    model_summaries,
+    glance_with_mod_name_slow(models),
+    pattern = map(models)
+  ),
+  # Get model predictions
+  tar_target(
+    model_predictions,
+    augment_with_mod_name_slow(models),
+    pattern = map(models)
+  )
+)
+```
+
+::: challenge
+## Increasing Resources
+
+Q: How would you modify your `_targets.R` if your targets needed 200GB of RAM?
+
+::: hint
+Check the arguments for [`crew_controller_slurm`](https://wlandau.github.io/crew.cluster/reference/crew_controller_slurm.html#arguments-1).
+:::
+::: solution
+```R
+tar_option_set(
+  controller = crew_controller_slurm(
+    workers = 3,
+    script_lines = "module load R",
+    # Added this
+    slurm_memory_gigabytes_per_cpu = 200,
+    slurm_cpus_per_task = 1
+  )
+)
+```
+:::
+:::
+
+## HPC Workers
+
+Despite what you might expect, `crew` does not submit one Slurm job for each target. 
+Instead, it uses persistent workers, meaning that you define a pool of workers when configuring the workflow.
+In our example above we used 3 workers.
+For each worker, `crew` submits a single Slurm job, and these workers will process multiple targets over their lifetime.
+
+We can verify that this has happened using `sacct`:
+
+```{bash}
+sacct
+```
+
+The upside of this approach is that we don't have to work out the minutae of how long each target takes to build, or what resources it needs.
+It also means that we don't submit a lot of jobs, making our Slurm usage more efficient and easy to monitor.
+
+The downside of this mechanism is that **the resources of the worker have to be sufficient to build each of your targets**.
```

Comment:
maybe you mean "the worker have to be sufficient to build all of your targets"? Maybe an example would be more helpful.

Reply:
I used to have an example here, but I converted the example into the following challenge block. Is this not a good approach?

Reply:
I just didn't understand that sentence that well. The example in the challenge is clear, but if I were in the learner's shoes, I would probably be "guessing" the answer to the challenge with the explanation alone.

Reply:
hence why changing "each target" to "all targets" might make more sense, as it expresses that all the targets need to be able to run within the crew's resources.